verification/src/solidity/compiler_fetcher.rs

```diff
@@ -0,0 +1,329 @@
+use crate::compiler::{fetcher::Fetcher, version::CompilerVersion};
+use async_trait::async_trait;
+use primitive_types::H256;
+use std::{
+    collections::HashMap,
+    fs::{File, OpenOptions},
+    io::ErrorKind,
+    os::unix::prelude::OpenOptionsExt,
+    path::{Path, PathBuf},
+};
+use thiserror::Error;
+use url::Url;
+
+mod json {
+    use primitive_types::H256;
+    use serde::Deserialize;
+    use url::Url;
+
+    use crate::compiler::version::CompilerVersion;
+
+    #[derive(Debug, Deserialize, PartialEq)]
+    pub struct List {
+        pub builds: Vec<CompilerInfo>,
+    }
+
+    #[derive(Debug, Deserialize, PartialEq)]
+    #[serde(rename_all = "camelCase")]
+    pub struct CompilerInfo {
+        pub path: DownloadPath,
+        #[serde(with = "serde_with::rust::display_fromstr")]
+        pub long_version: CompilerVersion,
+        pub sha256: H256,
+    }
+
+    #[derive(Debug, Deserialize, PartialEq)]
+    #[serde(untagged)]
+    pub enum DownloadPath {
+        Url(Url),
+        Filename(String),
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum ListError {
+    #[error("fetching list json returned error: {0}")]
+    ListJsonFetch(reqwest::Error),
+    #[error("cannot parse list json file: {0}")]
+    ParseListJson(reqwest::Error),
+    #[error("error parsing 'path' field: {0}")]
+    Path(url::ParseError),
+}
+
+#[derive(Default)]
+pub struct Releases(HashMap<CompilerVersion, CompilerInfo>);
+
+#[derive(Debug)]
+pub struct CompilerInfo {
+    pub url: Url,
+    pub sha256: H256,
+}
+
+impl Releases {
+    pub async fn fetch_from_url(compilers_list_url: &Url) -> Result<Self, ListError> {
+        let list_json_file: json::List = reqwest::get(compilers_list_url.to_string())
+            .await
+            .map_err(ListError::ListJsonFetch)?
+            .json()
+            .await
+            .map_err(ListError::ParseListJson)?;
+
+        Releases::try_from((list_json_file, compilers_list_url)).map_err(ListError::Path)
+    }
+}
+
+impl TryFrom<(json::List, &Url)> for Releases {
```

Comment:
May be make a struct/type alias, which will serve as more descriptive naming, like
```rust
pub struct UrledList(List, Url);
// Or?
pub type UrledList = (List, Url);

impl TryFrom<UrledList>...

// Usage
let releases: Releases = UrledList(list, url).into()
```

And same for the CompilerInfo

Reply:
urled list? sound like list of urls